### ISSUE TITLE
[Feature Proposal]: WCAG-accessible colour combinations generator

### ISSUE THREAD

**Issue by FrisovanDijk** [tailwindcss/tailwindcss#1171]
Before switching to Tailwind I used a custom generation built in SASS. One of the features I used a lot was the generation of WCAG-compliant contrast classes for text with background colours. Having it available, especially for mock-ups, removes mental overhead from the development process.

The generation was based on the perceived brightness of the background colour, with black or white as a text colour based on the WCAG guidelines. It would generate classes like `wcag-bg-blue-200`, resulting in black text on a white background. The idea is [further explained here](https://www.nbdtech.com/Blog/archive/2008/04/27/Calculating-the-Perceived-Brightness-of-a-Color.aspx).

The full calculations in SASS (can't find the original author):
````
// Calculates the brightness of a color
@function color-brightness($color) {
  $red-component: red($color);
  $green-component: green($color);
  $blue-component: blue($color);

  $red-magic-number: 241;
  $green-magic-number: 691;
  $blue-magic-number: 68;
  $brightness-divisor: $red-magic-number + $green-magic-number + $blue-magic-number;

  // Calculate a brightness value in 3d color space between 0 and 255
  $number: sqrt((($red-component * $red-component * $red-magic-number) + ($green-component * $green-component * $green-magic-number) + ($blue-component * $blue-component * $blue-magic-number)) / $brightness-divisor);

  // Convert to percentage and return
  @return 100% * $number / 255;
}

// Apply the correct WCAG 2.1 text color with high enough contrast
@function adaptive-text-color($background) {
  @if color-brightness($background) < 65% {
    @return #fff;
  } @else {
    @return #000;
  }
````

It seems like such a feature would fit with the spirit of Tailwind.

**Comment by tlgreg** [comment]
How would that work with tailwind's utility-first approach? Auto-generating a `text-for-bg-*` utilities? Then we would end up with a lot of redundant utilities though. Or just a tool to use in your config?

If you mean something to use in your config then you can use any node.js library out there to manipulate colors.
For example [color](https://github.com/Qix-/color) has WCAG helpers to determine if a color is dark or light:
```js
const Color = require('color')
const textColor = color => Color(color).isDark() ? '#fff' : '#000'
const badass = '#bada55'
const textOnBadass = textColor(badass) // => '#000'
```

**Comment by FrisovanDijk** [comment]
It's not a css, but an accessibility utility. Accessibility is no longer a thing you should do, but in many cases a legal requirement (up to a certain point). From my point-of-view, tools that make WCAG-compliance easier definitely have a place in a utility-first approach. Breaking from accessibility guidelines should be the exception.

As for redundancy, it promotes DRY if you work with lots of components. Even programmatic assignment of WCAG-compliant classes becomes easier when working with shades on a spectrum that is on the border of accessibility.

**Comment by adamwathan** [comment]
No plans to introduce any complex multi-property classes like this, if you want to extend Tailwind with this sort of thing yourself though of course you can by doing something like what @tlgreg is suggesting 👍

**Comment by craigerskine** [comment]
Here's my solution. I'm outputting my compiled Tailwind CSS as a `.less` file and then my final build process takes care of the color math. The big thing here is I'm disabling the default TW `backgroundColors` and `textColors` modules and replacing them with my own "plugins". I have to recreate the textColors plugin because I need to have the `.text-{color}` classes render AFTER the `.bg-{color}` classes so I can override them on elements that have `.bg-` classes if need be. It works amazingly well. Anyway, here is it:

````
modules: {
  ...
  backgroundColors: false, // we are using a custom plugin for this
  textColors: false, // we are using a custom plugin for this
  ...
},

plugins: [
  ...
  function({ addUtilities, e, variants }){
    const _ = require('lodash')
    // bg with color contrast
    const colorBgContrast = _.fromPairs(
      _.map(colors, (value, key) => {
        if(value === 'currentColor' || value === 'transparent'){
          return [
            `.${e(`bg-${key}`)}`,
            {
              'background-color': value,
            },
          ]
        } else {
          return [
            `.${e(`bg-${key}`)}`,
            {
              'background-color': value,
              'color': 'if((abs(luma('+ value +')) > 30), shade('+ value +', 80%), #FFF)',
            },
          ]
        }
      })
    )
    addUtilities(colorBgContrast, ['hover', 'focus'])

    // color after bg so it can be overridden via class stacking
    const colorFg = _.fromPairs(
      _.map(colors, (value, key) => {
        return [
          `.${e(`text-${key}`)}`,
          {
            'color': value
          },
        ]
      })
    )
    addUtilities(colorFg, ['hover', 'focus'])
  },
],
````
